rfcs/007-background-jobs.md

```diff
@@ -0,0 +1,350 @@
+describe how the framework is implemented. The intended audience is CouchDB
+developers using this framework to implement background jobs for indexing,
+replication, and couch-peruser.
+
+Both the job creation and the worker implementation APIs use a `JobOpts` map to
+represent a job. It MAY also contain these top level fields:
+
+  * `"priority"` : The value of this field will contain the `Priority` value of
+    the job. `Priority` is job-type specific.
+  * `"data"`: An opaque object (map), from the framework's point of view,
+    containing job-type specific data. It MAY contain an update sequence, or an
+    error message, for example.
+  * `"cancel"` : Boolean field defaulting to `false`. If `true` indicates the
+    user intends to stop a job's execution.
+  * `"resubmit"` : Boolean field defaulting to `false`. If `true` indicates
+    the job should be re-submitted.
+
+### Job Creation API ###
+
+```
+add(Type, JobId, JobOpts) -> ok | {error, Error}
+```
+ - Add a job to be executed by a background worker.
+
+```
+remove(Type, JobId) -> ok | not_found
+```
+ - Remove a job. If it is running, it will be stopped, then it will be removed
+   from the job table.
+
+```
+resubmit(Type, JobId) -> ok | not_found
+```
+ - Indicates that the job should be re-submitted for execution.
+
+```
+get_job(Type, JobId) -> {ok, JobOpts, JobState}
+```
+ - Return `JobOpts` and the `JobState`. `JobState` value MAY be:
+  * `"pending"` : This job is pending.
+  * `"running"` : This job is currently running.
+  * `"finished"` : This job has finished running and is not pending.
+
+### Worker Implementation API
+
+This API is to be used when implementing workers for various job types. The general pattern
+is to call `accept()` from something like a job manager, then for each accepted
+job spawn a worker to execute it, and then resume calling `accept()` to get
+other jobs. When a job is running, the worker MUST periodically call `update()`
+to prevent the activity monitor from re-enqueueing it. When the worker decides to stop
+running a job, they MUST call `finish()` to indicate that the job has finished running.
+
+```
+accept(Type[, MaxPriority]) -> {ok, JobId, WorkerLockId} | not_found
+```
+
+ - Dequeue a job from the pending queue. `WorkerLockId` is a UUID indicating
+   that the job is owned exclusively by the worker. `WorkerLockId` will be
+   passed as an argument to other API functions below and will be used to
+   verify that the current worker is still the only worker executing that job.
+   `MaxPriority` is an optional value which can limit the maximum job priority
+   that will be accepted. Jobs with priorities higher than that will not be
+   accepted. The intended usage is to allow `Priority` to be used as a way to
+   schedule job for executing at a future date. For example, `Priority` MAY
+   indicate that a replication job which has been repeatedly failing should
+   not execute any sooner than one hour from now.
```

Comment:
The concept of `Priority` was removed and is not `ScheduledTime` parameter. This makes it more clear how it will be used, is less abstract and easier to reason about.